### ISSUE TITLE
is this support hl target?

### ISSUE THREAD

**Issue by sonygod** [HaxeFoundation/record-macros#46]
is this support hl target?

**Comment by jonasmalacofilho** [comment]
Possibly, but needs to be tested.

**Comment by jonasmalacofilho** [comment]
Update: both SQLite and MySQL tests are currently failing.

```
$ haxe -D UTEST_PRINT_TESTS -D UTEST_PATTERN=Sqlite test-hl.hxml
$ hl build/test.hl
Running SqliteTest...
    testIssue34
    testIssue19
    testIssue6
    testData
    testDateQuery
SIGNAL 11
StringBuf.add(/usr/local/share/haxe/std/hl/_std/StringBuf.hx:58)
hl.types.ArrayObj.join(/usr/local/share/haxe/std/hl/types/ArrayObj.hx:46)
sys.db.$TableCreate.create(sys/db/TableCreate.hx:91)
CommonDatabaseTest.setup(CommonDatabaseTest.hx:40)
CommonDatabaseTest.~__initializeUtest__.0(CommonDatabaseTest.hx:27)
utest.ITestHandler.runSetup(utest/ITestHandler.hx:37)
utest.ITestHandler.execute(utest/ITestHandler.hx:32)
utest.Runner.runFixture(utest/Runner.hx:327)
utest._Runner.ITestRunner.runFixtures(utest/Runner.hx:411)
utest._Runner.ITestRunner.runCases(utest/Runner.hx:378)
utest._Runner.ITestRunner.run(utest/Runner.hx:358)
utest.Runner.run(utest/Runner.hx:274)
$Main.main(Main.hx:36)
fun$945(?:1)
[2]    158750 segmentation fault (core dumped)  hl build/test.hl
```

```
$ haxe -D UTEST_PRINT_TESTS -D UTEST_PATTERN=Mysql test-hl.hxml
$ hl build/test.hl mysql://[...]
Running MysqlTest...
    testIssue34
SIGNAL 11
CommonDatabaseTest.teardown(CommonDatabaseTest.hx:47)
CommonDatabaseTest.~__initializeUtest__.1(CommonDatabaseTest.hx:45)
utest.ITestHandler.runTeardown(utest/ITestHandler.hx:96)
utest.ITestHandler.runTest(utest/ITestHandler.hx:66)
utest.ITestHandler.checkSetup(utest/ITestHandler.hx:55)
utest.Async.then(utest/Async.hx:106)
utest.ITestHandler.runSetup(utest/ITestHandler.hx:47)
utest.ITestHandler.execute(utest/ITestHandler.hx:32)
utest.Runner.runFixture(utest/Runner.hx:327)
utest._Runner.ITestRunner.runFixtures(utest/Runner.hx:411)
utest._Runner.ITestRunner.runCases(utest/Runner.hx:378)
utest._Runner.ITestRunner.run(utest/Runner.hx:358)
utest.Runner.run(utest/Runner.hx:274)
$Main.main(Main.hx:36)
fun$945(?:1)
[2]    159016 segmentation fault (core dumped)  hl build/test.hl mysql://[...]
```

On the SQLite side it seems that the DB is closed too early.  There are also some issues with Dates.

Haven't started to look at what's causing the issues on MySQL.

**Comment by sonygod** [comment]
sorry to see that.

**Comment by emugel** [comment]
I would also be interested in having a working HL+Sqlite. 
Because when Unicode is needed, Neko+Sqlite is sometimes useless.

But so far I don't even know how to install what seems to be a missing hl lib, after running `hl myfile.hl`:
`src/module.c(337) : FATAL ERROR : Failed to load library sqlite.hdll`

**Comment by jonasmalacofilho** [comment]
If you're on Linux, that could have to do with HaxeFoundation/hashlink#323.

**Comment by emugel** [comment]
Yes on Linux. Now following the above (hashlink#323), thanks!

**Comment by jonasmalacofilho** [comment]
@emugel I'm been procrastinating finalizing that PR (missing: use system libsqlite on Linux, with both make and cmake).  This is because I know next to nothing about cmake, so it would be great if you could help me with that.

**Comment by emugel** [comment]
Forked https://github.com/jonasmalacofilho/hashlink/tree/sqlite-make-mysql-cmake-fixes branch. 

On Archlinux: 
* ran `make`, it built both `mysql.hdll` and `sqlite.hdll`
* ran `mv Makefile Makefile.old`, `rm *.hdll` then `cmake .` then `make`, it built them again. 

So I assume this works, and the only thing remaining to do is make this change suggested by Nicolas: 

> I think it should use the system installed libsqlite instead of the "include" one. ~~Also it should follow fmt that has its own include paths instead of adding itself to CFLAGS~~

I'm having a look at that then.

**Comment by emugel** [comment]
See https://github.com/jonasmalacofilho/hashlink/pull/1 for a first attempt

**Comment by emugel** [comment]
Something else. 
Whenever I run some `hl helloworld_sqlite.hl`, I get some `/home/john/code/haxe/hashlink/src/module.c(337) : FATAL ERROR : Failed to load library sqlite.hdll`. 

This is however not due to sqlite because I get the same with **fmt.hdll**: `/home/john/code/haxe/hashlink/src/module.c(337) : FATAL ERROR : Failed to load library fmt.hdll` whenever I run the hl of the following:

```haxe
/** This is to test whether hashlink has found fmt.hdll */
class T {
    public static function main() {
        var out = new hl.Bytes(1);
        var src = new hl.Bytes(1);
        hl.Format.digest(new hl.Bytes(1), new hl.Bytes(1), 1, /*algorithm:Int*/ 2);
    }
}
```

Giving `/home/john/code/haxe/hashlink/src/module.c(337) : FATAL ERROR : Failed to load library fmt.hdll`

For me `sudo make install` installs `hl` binary to `/usr/local/bin/hl` and the `*.hdll` files to `/usr/local/lib`. 

What appears is clear with `strace hl helloworld.hl` ([verbatim output](http://ix.io/24E5) ):

1. `/etc/ld.so.cache` is queried,
2. `/usr/local/lib/libhl.so.1` is found (because it appears in the cache (`ldconfig -p | grep libhl`)),
3. some dependency libs are also found,
4. Finally comes the moment of seeking`fmt.hdll`, where it tries various locations including `/usr/lib` but never `/usr/local/lib`. I show the last few lines: 

```
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 4
fstat(4, {st_mode=S_IFREG|0644, st_size=472312, ...}) = 0
mmap(NULL, 472312, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f40bfed6000
close(4)                                = 0
openat(AT_FDCWD, "/usr/lib/tls/x86_64/x86_64/fmt64.hdll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/tls/x86_64/x86_64", 0x7ffc8b4758f0) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/tls/x86_64/fmt64.hdll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/tls/x86_64", 0x7ffc8b4758f0) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/tls/x86_64/fmt64.hdll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/tls/x86_64", 0x7ffc8b4758f0) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/tls/fmt64.hdll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/tls", 0x7ffc8b4758f0)    = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/x86_64/x86_64/fmt64.hdll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/x86_64/x86_64", 0x7ffc8b4758f0) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/x86_64/fmt64.hdll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/x86_64", 0x7ffc8b4758f0) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/x86_64/fmt64.hdll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib/x86_64", 0x7ffc8b4758f0) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/fmt64.hdll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
stat("/usr/lib", {st_mode=S_IFDIR|0755, st_size=327680, ...}) = 0
munmap(0x7f40bfed6000, 472312)          = 0
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 4
fstat(4, {st_mode=S_IFREG|0644, st_size=472312, ...}) = 0
mmap(NULL, 472312, PROT_READ, MAP_PRIVATE, 4, 0) = 0x7f40bfed6000
close(4)                                = 0
openat(AT_FDCWD, "/usr/lib/fmt.hdll", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
munmap(0x7f40bfed6000, 472312)          = 0
fstat(1, {st_mode=S_IFREG|0644, st_size=9712, ...}) = 0
write(1, "/home/john/code/haxe/hashlink/sr"..., 96/home/john/code/haxe/hashlink/src/module.c(337) : FATAL ERROR : Failed to load library fmt.hdll
) = 96
exit_group(1)                           = ?
+++ exited with 1 +++
```
So seems it would work if I just move my `*.hdll` from `/usr/local/lib` to `/usr/lib`, but that'd be just a nasty one-time fix. 

Since `/usr/local/lib` appears in `/etc/ld.so.conf`, in `ldconfig -p`, that `sudo ldconfig` has been run, I think one possible cause is that files ending by `hdll` are not cached by ldconfig by default.

**Comment by emugel** [comment]
Update(s):

In my CMakeCache.txt last line is `prefix_result:INTERNAL=/usr/lib`. Most libs also have a hard coded `/usr/lib` path in this file. This contradicts the `/usr/local/lib` prefix that seems used for the installation.

**Comment by ncannasse** [comment]
@andyli

**Comment by jonasmalacofilho** [comment]
Should be fixed in the main branch (see #50).
